org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/apps/form/AddAuthorizationForm.java

```diff
@@ -0,0 +1,313 @@
+			confirmPanel.getButton(ConfirmPanel.A_OK).setEnabled(true);
+			trx.close();
+		}
+	}
+
+	private void processToken(String code, String trxName) throws IOException, GeneralSecurityException {
+		if (!Util.isEmpty(code)) {
+			if (code.startsWith("error=")) {
+				code = code.replace("error=", "");
+				if (code.equals("access_denied"))
+					FDialog.error(m_WindowNo, "Access Denied by User");
+				else
+					FDialog.error(m_WindowNo, code);
+				return;
+			}
+			int index = credentialCombo.getSelectedIndex();
+			MAuthorizationCredential credential = credentialList.get(index);
+			credential.set_TrxName(trxName);
+			String msg = credential.processToken(code, AEnv.getApplicationUrl());
+
+			this.detach();
+			
+			FDialog.info(m_WindowNo, this, msg);
+			return;
+		}
+		FDialog.error(m_WindowNo, "Request Fail or Aborted by User"); // TODO: translate
+	}
+
+	/**
+	 * Populate the credential combobox
+	 */
+	private void populateCredential() {
+		credentialList = MAuthorizationCredential.getCredentialsOfScope(parameter_scope);
+		for(MAuthorizationCredential credential : credentialList) {
+			credentialCombo.appendItem(credential.getName());
+		}
+		if (!credentialList.isEmpty()) {
+			credentialCombo.setSelectedIndex(0);
+			try {
+				buildClientListener(0);
+			} catch (Exception e) {
+				throw new RuntimeException(e);
+			}
+		}
+	}
+
+	/**
+	 * Build listener script
+	 * @param index
+	 */
+	private void buildClientListener(int index) {
+		MAuthorizationCredential credential = credentialList.get(index);
+		String url = credential.getAuthorizationRedirectURL_Full(AEnv.getApplicationUrl(), parameter_scope);
+
+		StringBuilder authScript = new StringBuilder("var x = window.outerWidth / 2 + window.screenX - (800 / 2);\n" + 
+			    "var y = window.outerHeight / 2 + window.screenY - (600 / 2);\n" + 
+			    "var authWindow = window.open('" + url + "','_blank','width=800, height=600, top='+y+', left='+x);\n");		
+		authScript.append("var timer = zk.Widget.$('#").append(timer.getUuid()).append("');console.log(timer);\n");
+		authScript.append("timer.windowRef = authWindow; timer.play();");		
+		confirmPanel.getButton(ConfirmPanel.A_OK).setWidgetListener("onClick", authScript.toString());
+		
+		buildTimerClientScript(credential);
+	}
+
+	/**
+	 * Confirm panel to process OK/Cancel buttons
+	 * @param evt
+	 */
+	private void onConfirmPanelAction(Event evt) {
+		if (evt.getTarget() == confirmPanel.getButton(ConfirmPanel.A_CANCEL)) {
+			this.detach();
+		} else if (evt.getTarget() == confirmPanel.getButton(ConfirmPanel.A_OK)) {
+			confirmPanel.getButton(ConfirmPanel.A_OK).setEnabled(false);
+			Clients.showBusy(null);
+		}
+	}
+
+	/**
+	 * Trigger action when highlighted
+	 */
+	@Override
+	public void doHighlighted() {
+		super.doHighlighted();
+		parameter_scope = Env.getContext(Env.getCtx(), getWindowNo(), "+SCOPE");
+		if (Util.isEmpty(parameter_scope)) {
+			parameter_scope = MAuthorizationAccount.AD_AUTHORIZATIONSCOPE_EMail;
```

Comment:
I think the logical things to do here if +SCOPE is missing is to provide a dropdown list of scope available for user to pick.

Reply:
Thanks @hengsin - I did the suggested change, if the +SCOPE is not passed, then a list of valid scopes is shown.